### PR TITLE
Add manual instrumentation stats/names under Live tab

### PR DESCRIPTION
**Pull request by akopich**
Now (in devmode only) all the manual instrumentation scopes
(sync and async) are shown under Live tab alongside with
the hooked functions.

Tests: Unit, Manual
Bug: http://b/228053666

### CODE REVIEW

**Comment by akopich** [comment]
http://screenshot/Aiy5WkbStYEPBxo

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 1]
```diff
@@ -94,15 +94,16 @@ std::string LiveFunctionsDataView::GetValue(int row, int column) {
...
-  const ScopeStats& stats = app_->GetCaptureData().GetScopeStatsOrDefault(function_id);
+  const uint64_t scope_id = GetScopeId(row);
+  const ScopeStats& stats = app_->GetCaptureData().GetScopeStatsOrDefault(scope_id);
 
-  const FunctionInfo& function = *GetFunctionInfoFromRow(row);
+  const FunctionInfo* function = GetFunctionInfoFromRow(row);
   switch (column) {
     case kColumnSelected:
-      return FunctionsDataView::BuildSelectedColumnsString(app_, function);
+      return function == nullptr ? ""
```
Comment (vickyliu-go4it) [inline src/DataViews/LiveFunctionsDataView.cpp:103]: Is it possible that the `function` is `nullptr` here?   
If not, I prefer to use `ORBIT_CHECK(function != nullptr)` before the `switch` instead of adding the `nullptr` in `kColumnSelected`, `kColumnModule` and `kColumnAddress` cases.

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 2]
```diff
@@ -200,14 +192,20 @@ void LiveFunctionsDataView::DoSort() {
...
-  const absl::flat_hash_map<uint64_t, FunctionInfo>& functions = functions_;
-
   switch (sorting_column_) {
     case kColumnSelected:
-      sorter = ORBIT_CUSTOM_FUNC_SORT(app_->IsFunctionSelected);
+      sorter = MakeFunctionSorter(
+          [this](const FunctionInfo& function_info) {
+            return app_->IsFunctionSelected(function_info);
+          },
+          ascending, false);
```
Comment (vickyliu-go4it) [inline src/DataViews/LiveFunctionsDataView.cpp:201]: Is there any problem with previous sorters? I prefer to not change them. 
Take the Selected column as example, the previous behavior in live functions data view and functions data view are the same. But I'm not sure whether they still keep consistent after the change.

Also the selected column not only shows whether a function is selected or not, but also shows the frame track is enabled or not. And now we are only sorting according to the selection.

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 3]
```diff
@@ -255,7 +255,7 @@ DataView::ActionStatus LiveFunctionsDataView::GetActionStatus(
       action == kMenuActionJumpToMin || action == kMenuActionJumpToMax) {
     if (selected_indices.size() != 1) return ActionStatus::kVisibleButDisabled;
 
-    uint64_t instrumented_function_id = GetInstrumentedFunctionId(selected_indices[0]);
+    uint64_t instrumented_function_id = GetScopeId(selected_indices[0]);
```
Comment (dpallotti) [inline src/DataViews/LiveFunctionsDataView.cpp:258]: A lot of the changes in this class assume that a scope_id is equal to the function_id for a function. This holds because of how `NameEqualityScopeIdProvider` is implemented, but it's a strong assumption, and it's hard to follow if you don't know that class. I feel like `AddFunction` should add to `functions_` by scope id now instead of by function_id.

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h [hunk 4]
```diff
@@ -14,6 +14,7 @@
 #include <utility>
 #include <vector>
 
+#include "../CompareAscendingOrDescending.h"
```
Comment (dpallotti) [inline src/DataViews/include/DataViews/LiveFunctionsDataView.h:17]: Fix this `#include`.

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 5]
```diff
@@ -116,9 +117,9 @@ std::string LiveFunctionsDataView::GetValue(int row, int column) {
     case kColumnStdDev:
       return orbit_display_formats::GetDisplayTime(absl::Nanoseconds(stats.ComputeStdDevNs()));
     case kColumnModule:
-      return function.module_path();
+      return function == nullptr ? "" : function->module_path();
     case kColumnAddress:
-      return absl::StrFormat("%#x", function.address());
+      return function == nullptr ? "" : absl::StrFormat("%#x", function->address());
```
Comment (dpallotti) [inline src/DataViews/LiveFunctionsDataView.cpp:122]: We collect a PC from manual instrumentation as well. Should we use it here at some point (if it's unique after aggregation)?

**Review by dpallotti — COMMENTED** [review]
Looking at the screenshot, I think we need some simple to tell whether a row is from a dynamically-instrumented function or from manual instrumentation. I guess you can use the fact that the last two columns are empty but that's not very clear. Is there any plan around this?

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 6]
```diff
@@ -94,15 +94,16 @@ std::string LiveFunctionsDataView::GetValue(int row, int column) {
     return "";
   }
 
-  const uint64_t function_id = GetInstrumentedFunctionId(row);
-  const ScopeStats& stats = app_->GetCaptureData().GetScopeStatsOrDefault(function_id);
+  const uint64_t scope_id = GetScopeId(row);
+  const ScopeStats& stats = app_->GetCaptureData().GetScopeStatsOrDefault(scope_id);
 
-  const FunctionInfo& function = *GetFunctionInfoFromRow(row);
+  const FunctionInfo* function = GetFunctionInfoFromRow(row);
```
Comment (florian-kuebler) [inline src/DataViews/LiveFunctionsDataView.cpp:100]: So you changed GetFunctionInfoFromRow to return a pointer, to express it being null on manual instrumentation, right? I'd prefer to make it even more explicit and use std::optional<FunctionInfo> here instead.

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 5 above)
Comment (florian-kuebler) [reply src/DataViews/LiveFunctionsDataView.cpp:122]: Same for the module (again, if it is unique).

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 3 above)
Comment (florian-kuebler) [reply src/DataViews/LiveFunctionsDataView.cpp:258]: Agree. The following code reads quite strange.

**Review by florian-kuebler — COMMENTED** [review]
Thanks for the change. Only minor additions to what have been said.

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 1 above)
Comment (vickyliu-go4it) [reply src/DataViews/LiveFunctionsDataView.cpp:103]: Never mind, just notice that this can be null on manual instrumentation.

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h (hunk 4 above)
Comment (akopich) [reply src/DataViews/include/DataViews/LiveFunctionsDataView.h:17]: What do you mean "fix" ?

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 5 above)
Comment (akopich) [reply src/DataViews/LiveFunctionsDataView.cpp:122]: I guess, that's for a subsequent PR, right?

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h (hunk 4 above)
Comment (dpallotti) [reply src/DataViews/include/DataViews/LiveFunctionsDataView.h:17]: Don't use `../`

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 6 above)
Comment (akopich) [reply src/DataViews/LiveFunctionsDataView.cpp:100]: You're right. I've given std::optional a thought. Would it be `std::optional<FunctionInfo>`? This would copy an instance of `FunctionInfo`. Or rather `std::optional<FunctionInfo&>`, that would effectively say "the reference is a non-null pointer, but we wrap it into optional, so it can be null", thus making its semantics extremely close to that of a pointer. I don't see the benefit as long as we treat std::optional in the same way we treat pointers.

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h (hunk 4 above)
Comment (akopich) [reply src/DataViews/include/DataViews/LiveFunctionsDataView.h:17]: Sorry, I don't get it. `CompareAscendingOrDescending.h` is in `src/DataViews`, while `LiveFunctionsDataView.h` is in `src/DataViews/include/DataViews`. How should I include it?

Comment (dpallotti) [reply src/DataViews/include/DataViews/LiveFunctionsDataView.h:17]: That tells you that something is wrong. You shouldn't include module-private headers from module-public headers. This means that `CompareAscendingOrDescending.h` needs to be moved to the `include` directory.

In addition, as `CompareAscendingOrDescending.h` was a private header and its content was supposed to be used only from inside the module, its content should be moved to the `orbit_data_views_internal` namespace.

Of course an alternative would be to not use that header at all from `LiveFunctionsDataView.h`, but you need it in a template so I don't know if that's possible.

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 6 above)
Comment (vickyliu-go4it) [reply src/DataViews/LiveFunctionsDataView.cpp:100]: There is similar discussions in go/totw/163.
I agree that `optional` here would be more readable, but the suggestion is to use pointer and let nullptr indicate "does not exist." 

I prefer to add a comments to the `GetFunctionInfoFromRow` method to explain when we return `nullptr` for this case.    
Previously `indices_` stores function ids for `FunctionInfo`s in `functions_`, this helps us to map the index / row in the view to the actual associated function.   
Now we change `indices_` to store scope id and it does not always have a matching `FunctionInfo` in the `functions_` if it a manually instrumented function.   
I would also like to add a comment above `indices_` and `functions_` to explain this a little bit.

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 5 above)
Comment (dpallotti) [reply src/DataViews/LiveFunctionsDataView.cpp:122]: > I guess, that's for a subsequent PR, right?

Yes yes (hence my "at some point").

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 6 above)
Comment (florian-kuebler) [reply src/DataViews/LiveFunctionsDataView.cpp:100]: I don't think that copying FunctionInfo here is really a big deal, which is why I would prefer the optional.
However, I agree that we should not use optional<FunctionInfo&>, and rather use the nullptr with a comment version then.

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 3 above)
Comment (akopich) [reply src/DataViews/LiveFunctionsDataView.cpp]: Good point. As discussed offline, I add FunctionIdToScopeId map (which is identity for now).

For now I keep function_id in a couple of places (action menu and csv export) as these will need to be changed anyway.

File: src/DataViews/include/DataViews/DataView.h [hunk 7]
```diff
@@ -189,6 +189,7 @@ class DataView {
   virtual void DoFilter() {}
   FilterCallback filter_callback_;
 
+  // contains a list of scope_id in the displayed order
```
Comment (vickyliu-go4it) [inline src/DataViews/include/DataViews/DataView.h:192]: "Contains"

File: src/DataViews/include/DataViews/CompareAscendingOrDescending.h [hunk 8]
```diff
@@ -2,16 +2,16 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-#ifndef DATA_VIEWS_COMPARE_ASCENDING_OR_DESCENDING_H_
-#define DATA_VIEWS_COMPARE_ASCENDING_OR_DESCENDING_H_
+#ifndef DATA_VIEWS_INTERNAL_COMPARE_ASCENDING_OR_DESCENDING_H_
```
Comment (vickyliu-go4it) [inline src/DataViews/include/DataViews/CompareAscendingOrDescending.h:5]: Should this be `DATA_VIEWS_INTERNAL_COMPARE_ASCENDING_OR_DESCENDING_H_`, as the format is 
`<PATH>_<FILE>_H_` ([go/cstyle#The__define_Guard](go/cstyle#The__define_Guard))

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 9]
```diff
@@ -326,12 +327,12 @@ DataView::ActionStatus LiveFunctionsDataView::GetActionStatus(
 
 void LiveFunctionsDataView::OnIteratorRequested(const std::vector<int>& selection) {
   for (int i : selection) {
-    uint64_t instrumented_function_id = GetInstrumentedFunctionId(i);
+    uint64_t scope_id = GetScopeId(i);
     const FunctionInfo* instrumented_function = GetFunctionInfoFromRow(i);
-    const ScopeStats& stats =
-        app_->GetCaptureData().GetScopeStatsOrDefault(instrumented_function_id);
+    ORBIT_CHECK(instrumented_function != nullptr);
```
Comment (vickyliu-go4it) [inline src/DataViews/LiveFunctionsDataView.cpp:332]: Why this cannot be a `nullptr`?

**Review by vickyliu-go4it — APPROVED** [review]
LGTM! Thanks for making this!

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 9 above)
Comment (akopich) [reply src/DataViews/LiveFunctionsDataView.cpp:332]: Currently, it can, but only in devmode. The check is added for the sake of a log message. I plan to make the action of adding an iterator inactive (and even invisible) for anything what is not a hooked function, so this will never be null.

Comment (vickyliu-go4it) [reply src/DataViews/LiveFunctionsDataView.cpp:332]: >  I plan to make the action of adding an iterator inactive (and even invisible) for anything what is not a hooked function

Do you have a bug for that? If so, you can add a TODO here.

File: src/ClientData/ScopeIdProvider.cpp [hunk 10]
```diff
@@ -69,6 +74,13 @@ uint64_t NameEqualityScopeIdProvider::ProvideId(const TimerInfo& timer_info) {
   return orbit_client_data::kInvalidScopeId;
 }
 
+[[nodiscard]] std::vector<uint64_t> NameEqualityScopeIdProvider::GetAllProvidedScopeIds() const {
```
Comment (dpallotti) [inline src/ClientData/ScopeIdProvider.cpp:77]: So this doesn't return scope_ids corresponding to functions? This seems wrong as this class also provides ids for function. Also, is that correct in `LiveFunctionsDataView`?

File: src/DataViews/include/DataViews/CompareAscendingOrDescending.h (hunk 8 above)
Comment (dpallotti) [reply src/DataViews/include/DataViews/CompareAscendingOrDescending.h:5]: "internal" is not part of the directory name. This should be `DATA_VIEWS_COMPARE_ASCENDING_OR_DESCENDING_H_` like before.

File: src/OrbitGl/App.cpp [hunk 11]
```diff
@@ -2417,7 +2417,7 @@ const InstrumentedFunction* OrbitApp::GetInstrumentedFunction(uint64_t function_
   return HasCaptureData() ? GetCaptureData().GetInstrumentedFunctionById(function_id) : nullptr;
 }
 
-void OrbitApp::SetVisibleFunctionIds(absl::flat_hash_set<uint64_t> visible_function_ids) {
+void OrbitApp::SetVisibleScopeIds(absl::flat_hash_set<uint64_t> visible_function_ids) {
   data_manager_->set_visible_function_ids(std::move(visible_function_ids));
```
Comment (dpallotti) [inline src/OrbitGl/App.cpp:2421]: This is renamed but the parameter is still called `visible_function_ids` and the implementation still calls `DataManager::set_visible_function_ids`. Which is it?

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h [hunk 12]
```diff
@@ -58,11 +59,12 @@ class LiveFunctionsDataView : public DataView {
...
   void DoFilter() override;
   void DoSort() override;
-  [[nodiscard]] uint64_t GetInstrumentedFunctionId(uint32_t row) const;
+  [[nodiscard]] uint64_t GetScopeId(uint32_t row) const;
   [[nodiscard]] std::optional<orbit_client_data::FunctionInfo>
   CreateFunctionInfoFromInstrumentedFunction(
       const orbit_grpc_protos::InstrumentedFunction& instrumented_function);
 
+  // Maps scope_ids corresponding to dynamically instrumented functions to FunctionInfo instances
   absl::flat_hash_map<uint64_t, orbit_client_data::FunctionInfo> functions_{};
```
Comment (dpallotti) [inline src/DataViews/include/DataViews/LiveFunctionsDataView.h:68]: I think that renaming to `scope_id_to_function_` or `functions_by_scope_id_` can help more than the comment.

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 13]
```diff
@@ -315,7 +316,7 @@ DataView::ActionStatus LiveFunctionsDataView::GetActionStatus(
   }
 
   for (int index : selected_indices) {
-    uint64_t instrumented_function_id = GetInstrumentedFunctionId(index);
+    uint64_t instrumented_function_id = GetScopeId(index);
```
Comment (dpallotti) [inline src/DataViews/LiveFunctionsDataView.cpp:319]: Is this a scope_id or a function_id?

File: src/ClientData/ScopeIdProvider.cpp (hunk 10 above)
Comment (akopich) [reply src/ClientData/ScopeIdProvider.cpp:77]: >So this doesn't return scope_ids corresponding to functions?

It does return all the provided ids, be it for hooked functions or not. So for now it's like `{1,2,3,4,5...}`

>So this doesn't return scope_ids corresponding to functions? This seems wrong as this class also provides ids for function. Also, is that correct in LiveFunctionsDataView?

In `LiveFunctionsDataView` I explicitly check if the id corresponds to a function or not.  Not sure what you mean, though.

File: src/OrbitGl/App.cpp (hunk 11 above)
Comment (akopich) [reply src/OrbitGl/App.cpp:2421]: Good point. These are all scope ids.

File: src/ClientData/ScopeIdProvider.cpp (hunk 10 above)
Comment (dpallotti) [reply src/ClientData/ScopeIdProvider.cpp:77]: Oooh I forgot that `scope_id_to_name_` is also initialized by `Create`... Ignore.

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h (hunk 12 above)
Comment (akopich) [reply src/DataViews/include/DataViews/LiveFunctionsDataView.h:68]: I'd still keep the comment. It points out, only the scope_ids corresponding to functions are present.

File: src/DataViews/LiveFunctionsDataView.cpp (hunk 13 above)
Comment (akopich) [reply src/DataViews/LiveFunctionsDataView.cpp:319]: `scope_id` it is. Anyway, I was about to rewrite the method in the next PR.

Comment (akopich) [reply src/DataViews/LiveFunctionsDataView.cpp:319]: As of now it also dereferences null.

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h (hunk 12 above)
Comment (dpallotti) [reply src/DataViews/include/DataViews/LiveFunctionsDataView.h]: Definitely, even better.

**Review by dpallotti — APPROVED** [review]

**Review by florian-kuebler — APPROVED** [review]
LGTM, thanks!
